feat: updated documentation of document creator with document storage (#10)
TradeTrust/documentation: src/document-creator/document-creator.mdx

### Config file structure

Below is an example of the configuration file.

```json
{
  "network": "ropsten",
  "wallet": "{\"address\":\"6a36c563a5350d7be66c801f901a67...\", ...}}",
  "forms": [{...}],
  "documentStorage": {
    "apiKey": "kNb15YYZ6N1zBlYd25cjj8PLgK6YAuvN9Gf7fPM1",
    "url": "https://api-ropsten.tradetrust.io/storage"
  }
}
```

Let's break down the required fields for this config file.

We will need:

- [Network field](#network-field)
- [Wallet field](#wallet-field)
- [Forms field](#forms-field)
- [Document Storage field](#document-storage-field)

---

## Network field

The `"network"` field is a string that refers to the network in which the document is created in.

As of now, we only cater to 3 networks.

- Mainnet : `"homestead"`
- Ropsten testnet : `"ropsten"`
- Rinkeby testnet : `"rinkeby"`

---

## Wallet field

The `"wallet"` field is a string that refers to your ethereum wallet.

If you have the wallet you can convert the wallet into a string by following these steps:

1. open your browser. (preferably mozella firefox)

2. open inspect element on any page.

3. go to the console tab.

4. type: "Node", and press enter.

> 5. type: "JSON.stringify({wallet})".

Reviewer: can just say

If you already have an existing wallet.json, the only thing you need to do is to run JSON.stringify over the contents in order to serialise it into a string.

This can be achieved using either the NodeJS console or the browser console, using the command `JSON.stringify({wallet})` where `{wallet}` is the contents of your wallet.json file.

For example: 

> w = { "address ": "6a36c563a5350d7be66c801f901a67 ... }
> JSON.stringify(w)

Author: ok

{
  "network": "ropsten",
  "wallet": "{\"address\":\"6a36c563a5350d7be66c801f901a67...\", ...}}",
  "forms": [{...}],
  "documentStorage": {
    "apiKey": "kNb15YYZ6N1zBlYd25cjj8PLgK6YAuvN9Gf7fPM1",
    "url": "https://api-ropsten.tradetrust.io/storage"
  }
}
```

Let's break down the required fields for this config file.

We will need:

- [Network field](#network-field)
- [Wallet field](#wallet-field)
- [Forms field](#forms-field)
- [Document Storage field](#document-storage-field)

## Network field

The `"network"` field is a string that refers to the network in which the document is created in.

As of now, we only cater to 3 networks.

- Mainnet : `"homestead"`
- Ropsten testnet : `"ropsten"`
- Rinkeby testnet : `"rinkeby"`

## Wallet field

> The `"wallet"` field is a string that refers to your etherum wallet.

Reviewer: ```suggestion
The `"wallet"` field is a string that refers to your ethereum wallet.
```

Author: ok

> If you have the wallet you can convert the wallet into a string like the above example, please use this [online tool](https://tools.knowledgewalls.com/jsontostring).

Reviewer: i think we shouldn't be recommending people to copy and paste their wallets into random online forms 🐱
can just suggest people to open their browser console and paste in the wallet object and do JSON.stringify there if necessary or modify oa cli to have a flag for stringified output

(why is it a stringified json instead of just an object?)

Author: ok will remove the online form.

regarding your question im not sure, need ask raymond. when i was doing it with him its already a stringified json, i thought it was like that all along.

> Alternatively, you can refer to [Open Attestation's documenation](https://github.com/Open-Attestation/open-attestation-cli#wallet) to create your wallet.(Note: you will need to have Open Attestation Cli installed)

Reviewer: ```suggestion
Alternatively, you can refer to [Open Attestation's documentation](https://github.com/Open-Attestation/open-attestation-cli#wallet) to create your wallet.(Note: you will need to have Open Attestation Cli installed)
```

Author: ok

Below is an example of the `"defaults"` field:

```json
"defaults": {
  "$template": {
    "type": "EMBEDDED_RENDERER",
    "name": "COVERING_LETTER",
    "url": "https://generic-templates.openattestation.com"
  },
  "issuers": [
    {
      "name": "Demo Issuer",
      "documentStore": "0x8bA63EAB43342AAc3AdBB4B827b68Cf4aAE5Caca",
      "identityProof": {
        "type": "DNS-TXT",
        "location": "demo-tradetrust.openattestation.com"
      }
    }
  ],
  "name": "Covering Letter",
  "logo": "https://www.aretese.com/images/govtech-animated-logo.gif",
  "title": "Documents Bundle",
  "remarks": "Some very important documents in here for some submission",
  "uiSchema": {
    "remarks": {
      "ui:widget": "textarea"
    }
  }
},
```

#### `"$template"` field

The `"$template"` field is an object that refers to the custom renderer, please refer to [creating document renderer](https://openattestation.com/docs/advanced/custom-renderer).

#### `"issuers"` field

The `"issuers"` is an array of the issuer object. The required fields for each issuer are `name`, `identityProof` and one of `documentStore` OR `tokenRegistry`.

> Transferrable document requires a [`tokenRegistry`](./token-registry), where as, a verifiable document requires a [`documentStore`](./deploying-store).

Reviewer: ```suggestion
Transferrable document requires a [`tokenRegistry`](./token-registry), whereas a verifiable document requires a [`documentStore`](./deploying-store).
```

Author: ok

If you want to accept all file types, remove the entire `"accept"` line from `"attachments"` object.

---

## Document Storage field

```json
"documentStorage": {

>   "apiKey": "kNb15YYZ6N1zBlYd25cjj8PLgK6YAuvN9Gf7fPM1",

Reviewer: Is this the actual ropsten api key ah? If it is let's just say we provide this endpoint for development purposes and there is no guarantee that the endpoint will remain the same plus there is "regular housekeeping" so files will get deleted after awhile - let's not say how long so that they don't rely on it

Author: ok, this is not the ropsten api key, since the ropsten endpoint doesn't need any api key right? 

This is just a random one i get from the oa function. should i change it to something else?

ok will add that this endpoint is for development purposes only. and that there is a regular housekeeping so files will get deleted after a while. also wont say how long.

  "url": "https://api-ropsten.tradetrust.io/storage"
}
```

> - Note: This document storage field is optional.

Reviewer: shall we explain what the purpose of the document storage is?
i.e QR code 

and also maybe can add a section at the end can show the format and link to the actions url ADR

Author: ok

> ##### If you do not want to upload your documents to a document storage enpoint you can omit this in the config file.

Reviewer: ```suggestion
##### If you do not want to upload your documents to a document storage endpoint you can omit this in the config file.
```

Author: ok

> To add a document storage endpoint, you will have to have the end point set up.

Reviewer: ```suggestion
To use a document storage endpoint, you will have to have the endpoint infrastructure already set up and configured.
```

Author: ok

You can refer to [infra template's storage section](/infra-templates#storage) for more information.

The `"documentStorage"` field is an object which expects an `"url"` field. `"apiKey"` can also be added.

### `"apikey"` field

> The `"apikey"` field is optional, it accepts a string as the api key.

Reviewer: ```suggestion
The `"apikey"` field is optional, it accepts a string as the API key.
```

Author: ok

> If your document storage endpoint does not requires any api key, you omit this field.

Reviewer: ```suggestion
If your document storage endpoint does not requires any API key, you should omit this field.
```

Author: ok

If you want to accept all file types, remove the entire `"accept"` line from `"attachments"` object.

---

## Document Storage field

```json
"documentStorage": {
  "apiKey": "kNb15YYZ6N1zBlYd25cjj8PLgK6YAuvN9Gf7fPM1",
  "url": "https://api-ropsten.tradetrust.io/storage"
}
```

_Note: This document storage field is optional._

##### If you do not want to upload your documents to a document storage endpoint you can omit this in the config file.

To use a document storage endpoint, you will have to have the endpoint infrastructure already set up and configured.

You can refer to [infra template's storage section](/infra-templates#storage) for more information.

The `"documentStorage"` field is an object which expects an `"url"` field. `"apiKey"` can also be added.

### `"apikey"` field

The `"apikey"` field is optional, it accepts a string as the API key.

> If your document storage endpoint does not requires any API key, you should omit this field.

Reviewer: ```suggestion
If your document storage endpoint does not require an API key, you should omit this field.
```

Author: ok

### `"url"` field

The `"url"` field accepts a string which will be the endpoint of the document storage.

_Note: please ensure that this document storage endpoint has a path `"/storage/queue"` that will return the queue number for storing your document in the document storage endpoint._

> Alternatively, for development purposes, we have set up a document storage endpoint.

Reviewer: ```suggestion
For the convenience of developers, we provide the below document storage endpoints for the Ropsten and Rinkeby testnets.
```

Author: ok

> This development purpose document storage endpoint currently is implemented for both ropsten testnet and rinkeby testnet only.

Reviewer: ```suggestion
```

Author: ok

### Ropsten document storage endpoint

```
"url": "https://api-ropsten.tradetrust.io/storage"

```

### Rinkeby document storage endpoint

```
"url": "https://api-rinkeby.tradetrust.io/storage"

```

> To use the default document storage endpoints, please use the correct url stated above for your network.

Reviewer: ```suggestion
These endpoints may only be used for valid and issued documents on the respective test networks, and may not be mixed. For example, a valid Ropsten document will fail validation and fail to be uploaded if the Rinkeby endpoint is specified.
```

Author: ok

> The API key is omitted for both these development endpoints.

Reviewer: ```suggestion
An API key is not required to be provided for these endpoints as they will accept unauthenticated requests.
```

Author: ok

> _Note: Because these document storage endpoints are just development purposes, there are some house keeping rules involved, uploaded documents will be deleted after some time._

Reviewer: ```suggestion
_Note: **The testnet endpoints do not retain documents permanently and any stored documents will be removed after a period of time**._
```

Author: ok
